_posts/2019-05-28-experience-and-opinions-developing-for-google-assistant-and-amazon-alexa.md

```diff
@@ -0,0 +1,79 @@
+
+On the other hand, Utah is a Google Drive Enterprise customer. Therefore, getting started with Google Assistant was very simple. No hoops to jump through.
+
+## Development {. text-left }
+
+### Debug Cycle
+
+I chose to use python for the Alexa skill. Partly because the [Alexa skills kit](https://developer.amazon.com/alexa-skills-kit) for [python](https://github.com/alexa/alexa-skills-kit-sdk-for-python) looked really nice when paired with [Flask](http://flask.pocoo.org/) and partly because it is nice to take a break from callbacks and promises once in a while.
+
+It was relatively easy to make Alexa send her requests to my [local flask server](https://github.com/agrc/digital-assisant-skills#development-ceremony) which was basically imitating a lambda function. This was a huge productivity boost as I was able to use the python debugger and all the tools I am already familiar with. [ngrok](https://dashboard.ngrok.com/get-started) makes it all possible by creating a temporary https url to my local machine that I could give to Alexa.
+
+The Google Assistant directs you to use Firebase functions for request fulfillment. Firebase only supports JavaScript for a node environment -- so JavaScript it was!
+
+I never set up a local server for the Google Assistant. While the Firebase CLI is awesome and you can start a local server, debugging what is happening in that local server is not simple. I could have set up an express server equivalent to flask but I chose not to. In hindsight, I would recommend the express approach since I frustratingly ended up `console.log` debugging in the Firebase logs.
+
+### Skills as Code
+
+I think it is a big win to have a skills training and design documented along with the implementation. Alexa knocks this out of the park with its [skill.json](https://github.com/agrc/digital-assisant-skills/blob/alexa-skill/skill.template.json) and [models](https://github.com/agrc/digital-assisant-skills/blob/alexa-skill/models/en-US.json). Developers can make changes without having to visit a website and deal with UI and UX. The ask CLI from Amazon is awesome and lets you do this and much more. You can deploy your function, update your skill and model, and test your skill. Adding a new question and training it is as simple as editing a json file. Sharing a demo or some code? Everything is documented as text.
+
+[Dialogflow](https://dialogflow.com/) is Google's machine learning natural language processing Alexa equivalent. Currently there is an API for Dialogflow but no supported clients. Everything you want to do has to go through the Dialogflow web console. The models are shared as a zip file that you have to export and import. The content within them is complicated and risky to edit by hand. I did not like having to import a zip file to look at how a demo was implemented.
```

Comment:
It'd be nice to add something about collaboration and source control benefits to skill.json and alexa models. Harder to keep things straight in dialogflow